shell.go

```diff
@@ -84,6 +92,94 @@ func NewShellWithClient(url string, c *gohttp.Client) *Shell {
 	}
 }
 
+// Load all shell urls with error checking
+func NewShellWithCheck(urls ...string) (*Shell, error) {
+	encountered := map[string]struct{}{}
+	for _, url := range urls {
+		if _, ok := encountered[url]; !ok {
+			encountered[url] = struct{}{}
+			sh := NewShell(url)
+			_, _, err := sh.Version()
+			if err == nil {
+				return sh, nil
+			}
+		}
+	}
+	return nil, errors.New(fmt.Sprintf("No working server in %v", urls))
+}
+
+func NewShellFromDefaultGateway() (*Shell, error) {
+	return NewShellWithCheck(DefaultGateway)
+}
+
+// Get all shell urls from api files
+func newShellFromAPIFiles(files ...string) (urls []string) {
+	for _, apifile := range files {
+		if data, err := ioutil.ReadFile(apifile); err == nil {
+			url := strings.Trim(string(data), "\n\t ")
+			urls = append(urls, url)
+		}
+	}
+	return
+}
+
+func NewShellFromAPIFiles(files ...string) (*Shell, error) {
+	return NewShellWithCheck(newShellFromAPIFiles(files...)...)
+}
+
+// Get api url from ~/.ipfs/api
+func newShellFromDefaultAPIFile() (urls []string) {
+	apifile, err := homedir.Expand(path.Join(DefaultPathRoot, DefaultApiFile))
+	if err != nil {
+		return urls
+	}
+	return newShellFromAPIFiles(apifile)
+}
+
+func NewShellFromDefaultAPIFile() (*Shell, error) {
+	return NewShellWithCheck(newShellFromDefaultAPIFile()...)
+}
+
+// get all shell urls from environmental variable IPFS_API and IPFS_PATH
+func newShellFromEnv() (urls []string) {
+	if ipfsAPI := os.Getenv("IPFS_API"); ipfsAPI != "" {
+		urls = append(urls, ipfsAPI)
+	}
+	if ipfsPath := os.Getenv("IPFS_PATH"); ipfsPath != "" {
+		apifile := path.Join(ipfsPath, DefaultApiFile)
+		urls = append(urls, newShellFromAPIFiles(apifile)...)
+	}
+	return
+}
+
+func NewShellFromEnv() (*Shell, error) {
+	return NewShellWithCheck(newShellFromEnv()...)
+}
+
+// Load working default shell address once
+func defaultShell() {
+	fmt.Println("ran defaultShell")
+	var urls []string
+	urls = append(urls, newShellFromEnv()...)
+	urls = append(urls, newShellFromDefaultAPIFile()...)
+	urls = append(urls, DefaultAPIAddr, DefaultGateway)
+	if sh, err := NewShellWithCheck(urls...); err == nil {
+		defaultShellAddr = sh.url
+	}
+}
+
+func DefaultShell() *Shell {
+	// Load once and cache working default shell address
+	defaultShellAddrLoadOnce.Do(defaultShell)
+	return NewShell(defaultShellAddr)
```

Comment:
If the apifile does exist, but an error occurred while reading the apifile, should we just return `nil`? This will also suprise users, as people will see unexpected panic. In this extreme case, should we just call panic?

Reply:
Hm. Yeah, that's unfortunate. We shouldn't panic. We could:

1. Return nil.
2. Return an error (stashing it somewhere). That makes the API a bit more difficult to use but that's probably not that much of an issue.
3. Return a special shell that always fails? Probably a bad idea.

I'd prefer 2.

Reply:
Do you mind clarify where should I put the error?

Reply:
Put it in a variable, guarded by a `sync.Once`.

Reply:
I did not add a `sync.Once` to reading api file. For the time being, reading api file will only be called by `NewLocalShell`, which is already guarded by a `sync.Once`. Sorry for the delay. I had a really busy week on some earthly bussiness.

Reply:
No need to apologize for delays, there's no rush on this.